src/csharp/Grpc.Core/Interceptors/CallInvokerExtensions.cs

```diff
@@ -0,0 +1,171 @@
+#region Copyright notice and license
+
+// Copyright 2018 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Linq;
+using Grpc.Core.Utils;
+
+namespace Grpc.Core.Interceptors
+{
+    /// <summary>
+    /// Extends the CallInvoker class to provide the interceptor facility on the client side.
+    /// This is an EXPERIMENTAL API.
+    /// </summary>
+    public static class CallInvokerExtensions
+    {
+        /// <summary>
+        /// Decorates an underlying <see cref="Grpc.Core.CallInvoker" /> to
+        /// intercept calls through a given interceptor.
+        /// </summary>
+        private class InterceptingCallInvoker : CallInvoker
+        {
+            readonly CallInvoker invoker;
+            readonly Interceptor interceptor;
+
+            /// <summary>
+            /// Creates a new instance of <see cref="Grpc.Core.Interceptors.CallInvokerExtensions.InterceptingCallInvoker" />
+            /// with the given underlying invoker and interceptor instances.
+            /// </summary>
+            public InterceptingCallInvoker(CallInvoker invoker, Interceptor interceptor)
+            {
+                this.invoker = GrpcPreconditions.CheckNotNull(invoker, "invoker");
+                this.interceptor = GrpcPreconditions.CheckNotNull(interceptor, "interceptor");
+            }
+
+            /// <summary>
+            /// Intercepts a simple blocking call with the registered interceptor.
+            /// </summary>
+            public override TResponse BlockingUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
+            {
+                return interceptor.BlockingUnaryCall(
+                    request,
+                    new ClientInterceptorContext<TRequest, TResponse>(method, host, options),
```

Comment:
btw, for high througput scenarios, allocating a new context  and a new lambda `(req, ctx) =>`  might be a significant performance hit as  extra allocations put additional load on GC (which seems to be the bottleneck according to my [not fully conclusive] observations). Not sure if we can do anything about this, here, so basically just FYI.

You also might consider making ClientInterceptorContext a struct (but that might have other challenges).

Reply:
Thinking about it, making it a struct is probably a better idea. It should be small enough and it is easier to avoid perf issues with the struct than a class down the road.